File: backend/nestjs/nlayer-architecture/minimal/src/auth/auth.controller.ts
import {
  Controller,
  Post,
  Get,
  Body,
  HttpCode,
  HttpStatus,
  UseGuards,
  Request,
} from '@nestjs/common';
import { AuthGuard } from '@nestjs/passport';
import { ApiTags, ApiOperation, ApiResponse, ApiBearerAuth } from '@nestjs/swagger';
import { AuthService } from './auth.service';
import { ResponseMessage } from '../common/decorators/response-message.decorator';
import { IsString } from 'class-validator';

class ValidateTokenDto {
  @IsString() token: string;
}

@ApiTags('Auth')
@Controller('auth')
export class AuthController {
  constructor(private readonly authService: AuthService) {}

  @Post('validate-token')
  @HttpCode(HttpStatus.OK)
  @ApiOperation({ summary: 'Validate external token' })
  @ApiResponse({ status: 200, description: 'Token validated' })
  @ApiResponse({ status: 401, description: 'Invalid or expired external token' })
  @ResponseMessage('Token validated')
  async validateToken(@Body() dto: ValidateTokenDto) {
    return this.authService.validateToken(dto.token);
  }

  @Get('me')
  @UseGuards(AuthGuard('jwt'))
  @ApiBearerAuth('JWT-auth')
  @ApiOperation({ summary: 'Get current user info from JWT claims' })
  @ApiResponse({ status: 200, description: 'Current user info' })
  @ResponseMessage('User info retrieved')
  async me(@Request() req: any) {
    return this.authService.getMe(req.user);
  }
}

File: backend/spring/nlayer-architecture/minimal/src/main/java/com/apptemplate/api/controller/AuthController.java
package com.apptemplate.api.controller;

import com.apptemplate.api.dto.ApiResponse;
import com.apptemplate.api.dto.AuthResponse;
import com.apptemplate.api.dto.RefreshTokenRequest;
import com.apptemplate.api.dto.UpdateProfileRequest;
import com.apptemplate.api.dto.ValidateTokenRequest;
import com.apptemplate.api.service.AuthService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/auth")
@RequiredArgsConstructor
@Tag(name = "Authentication", description = "Authentication and authorization endpoints")
public class AuthController {

    private final AuthService authService;

    @PostMapping("/validate-token")
    @Operation(summary = "Validate token", description = "Validate external token and receive internal JWT tokens")
    public ResponseEntity<ApiResponse<AuthResponse>> validateToken(@Valid @RequestBody ValidateTokenRequest request) {
        AuthResponse response = authService.validateToken(request);
        return ResponseEntity.ok(ApiResponse.success(response, "Token validated successfully"));
    }

    @PostMapping("/refresh")
    @Operation(summary = "Refresh token", description = "Get a new access token using refresh token")
    public ResponseEntity<ApiResponse<AuthResponse>> refresh(@Valid @RequestBody RefreshTokenRequest request) {
        AuthResponse response = authService.refresh(request);
        return ResponseEntity.ok(ApiResponse.success(response, "Token refreshed successfully"));
    }

    @PostMapping("/logout")
    @Operation(summary = "Logout", description = "Invalidate refresh token")
    public ResponseEntity<Void> logout(@RequestBody(required = false) RefreshTokenRequest request) {
        authService.logout(request != null ? request.getRefreshToken() : null);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/me")
    @Operation(summary = "Get current user", description = "Get basic info of authenticated user from JWT claims")
    public ResponseEntity<ApiResponse<AuthResponse.UserDto>> getCurrentUser() {
        AuthResponse.UserDto user = authService.getCurrentUser();
        return ResponseEntity.ok(ApiResponse.success(user, "User retrieved successfully"));
    }

    @GetMapping("/profile")
    @Operation(summary = "Get profile", description = "Get full profile of authenticated user from DB")
    public ResponseEntity<ApiResponse<AuthResponse.UserDto>> getProfile() {
        AuthResponse.UserDto profile = authService.getProfile();
        return ResponseEntity.ok(ApiResponse.success(profile, "Profile retrieved successfully"));
    }

    @PutMapping("/profile")
    @Operation(summary = "Update profile", description = "Update profile of authenticated user")
    public ResponseEntity<ApiResponse<AuthResponse.UserDto>> updateProfile(
            @Valid @RequestBody UpdateProfileRequest request) {
        AuthResponse.UserDto profile = authService.updateProfile(request);
        return ResponseEntity.ok(ApiResponse.success(profile, "Profile updated successfully"));
    }
}
